src/brod_client.erl

```diff
@@ -420,19 +420,36 @@ shutdown_pid(_) ->
         {ok, pid()} | {error, get_worker_error()}.
 get_partition_worker(ClientPid, Key) when is_pid(ClientPid) ->
   case erlang:process_info(ClientPid, registered_name) of
-    {registered_name, ClientId} when is_atom(ClientId) ->
-      get_partition_worker(ClientId, Key);
+    {registered_name, ClientId} -> get_partition_worker(ClientId, Key);
+
     _ ->
       %% This is a client process started without registered name
       %% have to call the process to get the producer/consumer worker
       %% process registration table.
       case safe_gen_call(ClientPid, get_workers_table, infinity) of
-        {ok, Ets}       -> lookup_partition_worker(ClientPid, Ets, Key);
+        {ok, Ets} -> lookup_partition_worker(ClientPid, Ets, Key);
         {error, Reason} -> {error, Reason}
       end
   end;
+
 get_partition_worker(ClientId, Key) when is_atom(ClientId) ->
-  lookup_partition_worker(ClientId, ?ETS(ClientId), Key).
+  case lookup_partition_worker(ClientId, ?ETS(ClientId), Key) of
+    {ok, Pid} ->
+      %% This is an attempt to help with a race condition
+      %% between ets lookups and what is being supervised.
+      %% If the worker process is returned form ets,
+      %% but it is not alive then there must be
+      %% an inflight worker deregistration request.
+      case is_process_alive(Pid) of
+        true -> {ok, Pid};
+        false -> partition_worker_error(Key)
```

Comment:
If I call `get_workers_table` then by that time the actual client has terminated.

Reply:
It is true when client is shutting down.
But it could also be that just one producer worker is restarting?
So I think it’s better to call get_workers_table
on gen call `exit` exception: return client_down
Otherwise return the (possibly new) pid from ets lookup.